util/tracing/tracer.go

```diff
@@ -0,0 +1,642 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"fmt"
+	"math/rand"
+	"regexp"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"context"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"golang.org/x/net/trace"
+)
+
+const maxLogsPerSpan = 1000
+
+// These constants are used to form keys to represent tracing context
+// information in carriers supporting opentracing.HTTPHeaders format.
+const (
+	prefixTracerState = "tidb-tracer-"
+	prefixBaggage     = "tidb-baggage-"
+	// prefixShadow is prepended to the keys for the context of the shadow tracer
+	// (e.g. LightStep).
+	prefixShadow = "tidb-shadow-"
+
+	fieldNameTraceID = prefixTracerState + "traceid"
+	fieldNameSpanID  = prefixTracerState + "spanid"
+	// fieldNameShadow is the name of the shadow tracer.
+	fieldNameShadowType = prefixTracerState + "shadowtype"
+
+	TiDBQueryLevelTracing = "tidb_query_level_trace"
+)
+
+// Tracer is our own custom implementation of opentracing.Tracer. It supports:
+//
+//  * forwarding events to x/net/trace instances
+//
+//  * recording traces. It has two types: query-level tracing and session-level tracing.
+//
+//  * lightstep traces. This is implemented by maintaining a "shadow" lightstep
+//    span inside each of our spans.
+//
+// Even when tracing is disabled, we still use this Tracer (with x/net/trace and
+// lightstep disabled) because of its recording capability (query-level
+// tracing needs to work in all cases).
+//
+// Tracer is currently stateless so we could have a single instance and it is currently careated
+// at final stage of creating a tidb-server.
+type Tracer struct {
+	// Pre-allocated noopSpan, used to avoid creating spans when we are not using
+	// x/net/trace or lightstep and we are not recording.
+	noopSpan noopSpan
```

Comment:
It seem can move `noopSpan` out, and  be a global variable, just classical singleton pattern, no need allocate for each `Tracer` instance?

then we can optimize other codes from type-assert cost

```
if _, noop := s.(*noopSpan); noop {
      ......
}
```
to  compare reference address cost

```
if  noop == _noopSpan_ {
	......
}
```